<a href="https://colab.research.google.com/github/roblei007/Aula-Projeto1/blob/L%C3%B3gica-de-Programa%C3%A7%C3%A3o-em-Python/C%C3%B3pia_de_aula_python_listas_dict_funcoes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🐍 Aula de Python - Parte 2: Listas, Dicionários e Funções

Bem-vindo(a) à segunda parte! Agora que você já domina variáveis, operadores, f-strings, while, for e range, vamos dar um salto de qualidade.

**Conteúdo desta aula:**

1. **Listas** — criar, acessar, adicionar, remover, buscar, ordenar
2. **Dicionários** — estrutura chave/valor, percorrer, modificar
3. **Funções** — criar seus próprios "mini-programas" reutilizáveis

Cada subtópico tem **4 exemplos práticos** e **1 exercício** para você praticar.

> 💡 **Dica:** Para executar uma célula, clique nela e pressione `Shift + Enter`.

---
# 📦 PARTE 1 — LISTAS

Listas são **coleções ordenadas de itens**. Imagine uma lista de compras, uma lista de alunos, ou uma lista de temperaturas — todos esses dados podem ser guardados em uma única variável do tipo **lista**.

**Características das listas:**
- São criadas com **colchetes** `[ ]`
- Os itens são separados por **vírgula**
- Podem conter **qualquer tipo de dado** (números, textos, até outras listas!)
- São **ordenadas**: cada item tem uma posição (índice) que começa em **0**
- São **mutáveis**: podemos adicionar, remover e modificar itens

```python
nomes = [Ana", "Bruno", "Carlos"]
numeros = [10, 25, 38, 47]
mistura = ["Python", 2025, 3.14, True]
```

## 📌 1.1. Criando e acessando listas

Cada item de uma lista tem uma **posição (índice)**. Para acessar um item, usamos `lista[índice]`.

**Atenção:** os índices começam em **0**, não em 1!

```
Lista:     ["Ana", "Bruno", "Carlos", "Diana"]
Índices:      0       1        2         3
Índ. neg.:   -4      -3       -2        -1   ← índices negativos contam do fim pro começo!
```

In [ ]:
# ============================================================
# EXEMPLO 1 - Criando uma lista e acessando itens pelo índice
# ============================================================

# Criando uma lista de frutas
# Cada item ocupa uma posição (índice) que começa em 0
frutas = ["maçã", "banana", "uva", "laranja", "manga"]

# Acessando cada item pelo seu índice
print(frutas[0])   # maçã    ← primeiro item (índice 0)
print(frutas[1])   # banana  ← segundo item
print(frutas[2])   # uva
print(frutas[3])   # laranja
print(frutas[4])   # manga   ← quinto item (índice 4)

# A função len() retorna quantos itens a lista tem
print(f"\nTotal de frutas: {len(frutas)}")

maçã
banana
uva
laranja
manga

Total de frutas: 5


In [ ]:
# ============================================================
# EXEMPLO 2 - Índices negativos (contando de trás pra frente)
# ============================================================

# Podemos usar índices NEGATIVOS para acessar itens de trás pra frente
# -1 = último item, -2 = penúltimo, e assim por diante

frutas = ["maçã", "banana", "uva", "laranja", "manga"]

print(frutas[-1])   # manga    ← último item
print(frutas[-2])   # laranja  ← penúltimo
print(frutas[-3])   # uva

# Isso é MUITO útil quando você não sabe o tamanho da lista
# e quer pegar o último item sem precisar calcular len(lista) - 1
print(f"\nO último da fila é: {frutas[-1]}")
print(f"O primeiro da fila é: {frutas[0]}")

manga
laranja
uva

O último da fila é: manga
O primeiro da fila é: maçã


In [ ]:
# ============================================================
# EXEMPLO 3 - Modificando itens e listas com tipos diferentes
# ============================================================

# Podemos MODIFICAR um item usando o índice
frutas = ["maçã", "banana", "uva"]
print("Antes:", frutas)

# Trocando 'banana' por 'morango'
frutas[1] = "morango"   # Substitui o item na posição 1
print("Depois:", frutas)

# Listas podem ter tipos DIFERENTES de dados misturados
# (embora o mais comum seja usar um tipo só)
dados_pessoa = ["Maria", 25, 1.68, True]   # nome, idade, altura, é_ativa
print(f"\nNome: {dados_pessoa[0]}")
print(f"Idade: {dados_pessoa[1]}")
print(f"Altura: {dados_pessoa[2]}")
print(f"Ativa? {dados_pessoa[3]}")

Antes: ['maçã', 'banana', 'uva']
Depois: ['maçã', 'morango', 'uva']

Nome: Maria
Idade: 25
Altura: 1.68
Ativa? True


In [ ]:
# ============================================================
# EXEMPLO 4 - Fatiamento (slicing) - pegando 'pedaços' da lista
# ============================================================

# Podemos pegar um 'pedaço' da lista usando lista[inicio:fim]
# Importante: o índice 'fim' NÃO é incluído (igual no range)

numeros = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
#  Índices: 0   1   2   3   4   5   6   7   8   9

# Pegando do índice 2 até o 4 (NÃO inclui o 5)
print(numeros[2:5])     # [30, 40, 50]

# Do começo até o índice 3 (exclusivo)
print(numeros[:3])      # [10, 20, 30]

# Do índice 6 até o final
print(numeros[6:])      # [70, 80, 90, 100]

# Os três últimos itens (índices negativos também funcionam!)
print(numeros[-3:])     # [80, 90, 100]

# Cópia completa da lista
print(numeros[:])       # [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

[30, 40, 50]
[10, 20, 30]
[70, 80, 90, 100]
[80, 90, 100]
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]


### ✏️ Exercício 1.1 — Criando e acessando listas

Você é DJ de uma festa e tem uma playlist com 6 músicas:

```python
playlist = ["Imagine", "Bohemian Rhapsody", "Hotel California", "Stairway to Heaven", "Thriller", "Billie Jean"]
```

Faça o seguinte:

1. Mostre a **primeira música** da playlist.
2. Mostre a **última música** usando índice negativo.
3. Mostre as **3 primeiras músicas** usando fatiamento.
4. Mostre as **3 últimas músicas** usando fatiamento com índice negativo.
5. Troque a terceira música por `"Let It Be"`.
6. Mostre o **tamanho total** da playlist usando `len()`.

In [ ]:
# 👉 Escreva seu código aqui:

playlist = ["Imagine", "Bohemian Rhapsody", "Hotel California", "Stairway to Heaven", "Thriller", "Billie Jean"]

# 1. Primeira música
print(playlist[0])

# 2. Última música (índice negativo)
print(playlist[-1])

# 3. Três primeiras músicas (fatiamento)
print(playlist[:3])

# 4. Três últimas músicas (fatiamento negativo)
print(playlist[-3:])

# 5. Trocar a terceira música
playlist[2] = "Let It Be"
print(playlist)

# 6. Tamanho da playlist
print(len(playlist))


Imagine
Billie Jean
['Imagine', 'Bohemian Rhapsody', 'Hotel California']
['Stairway to Heaven', 'Thriller', 'Billie Jean']
['Imagine', 'Bohemian Rhapsody', 'Let It Be', 'Stairway to Heaven', 'Thriller', 'Billie Jean']
6


## 📌 1.2. Adicionando elementos

Existem três formas principais de adicionar itens numa lista:

| Método | O que faz | Exemplo |
|--------|-----------|---------|
| `.append(item)` | Adiciona no **FINAL** da lista | `lista.append("novo")` |
| `.insert(índice, item)` | Adiciona em uma **POSIÇÃO ESPECÍFICA** | `lista.insert(0, "novo")` |
| `.extend(outra_lista)` | Adiciona **VÁRIOS ITENS** de uma vez | `lista.extend(["a", "b"])` |

In [ ]:
# ============================================================
# EXEMPLO 1 - Método .append() — adiciona no FINAL
# ============================================================

# Começamos com uma lista de compras vazia
compras = []

# .append() adiciona UM item no final da lista
compras.append("pão")
compras.append("leite")
compras.append("ovos")
compras.append("café")

# Mostra a lista completa
print("Lista de compras:", compras)

# Podemos verificar o tamanho a qualquer momento
print(f"Total de itens: {len(compras)}")

Lista de compras: ['pão', 'leite', 'ovos', 'café']
Total de itens: 4


In [ ]:
# ============================================================
# EXEMPLO 2 - Método .insert() — adiciona em POSIÇÃO específica
# ============================================================

# Lista de pessoas na fila do banco
fila = ["Ana", "Bruno", "Carlos"]
print("Fila inicial:", fila)

# .insert(índice, item) coloca o item NA posição indicada
# Os itens que já estavam ali são 'empurrados' para frente

# Adicionar alguém no INÍCIO da fila (prioridade)
fila.insert(0, "Dona Maria (preferencial)")
print("Após preferencial:", fila)

# Adicionar alguém na posição 2 (meio da fila)
fila.insert(2, "Pedro")
print("Após Pedro entrar no meio:", fila)

Fila inicial: ['Ana', 'Bruno', 'Carlos']
Após preferencial: ['Dona Maria (preferencial)', 'Ana', 'Bruno', 'Carlos']
Após Pedro entrar no meio: ['Dona Maria (preferencial)', 'Ana', 'Pedro', 'Bruno', 'Carlos']


In [ ]:
# ============================================================
# EXEMPLO 3 - Método .extend() — adiciona VÁRIOS de uma vez
# ============================================================

# Lista com alunos da turma A
turma_a = ["Ana", "Bruno", "Carlos"]

# Lista com alunos da turma B
turma_b = ["Diana", "Eduardo", "Fernanda"]

# .extend() junta os itens de uma lista no final da outra
turma_a.extend(turma_b)

print("Turma unificada:", turma_a)

# CUIDADO: .extend() é diferente de .append()
# Se usássemos .append(turma_b), a lista inteira viraria UM ITEM só!
turma_c = ["Gabriel", "Helena"]
turma_errada = ["Ana", "Bruno"]
turma_errada.append(turma_c)   # Errado! Coloca a LISTA como um item
print("\nCom append (errado):", turma_errada)   # ["Ana", "Bruno", ["Gabriel", "Helena"]]

Turma unificada: ['Ana', 'Bruno', 'Carlos', 'Diana', 'Eduardo', 'Fernanda']

Com append (errado): ['Ana', 'Bruno', ['Gabriel', 'Helena']]


In [ ]:
# ============================================================
# EXEMPLO 4 - Combinando com input() e while (cadastro dinâmico)
# ============================================================

# Uma lista pode ser preenchida pelo usuário dinamicamente!
# Vamos cadastrar nomes até o usuário digitar 'sair'

convidados = []   # Começamos com lista vazia

print("=== Cadastro de convidados ===")
print("(digite 'sair' para parar)\n")

while True:   # Loop que só para com 'break'
    nome = input("Nome do convidado: ")

    # Se o usuário digitou 'sair', sai do loop
    # .lower() converte para minúsculo, evita problema com 'SAIR' ou 'Sair'
    if nome.lower() == "sair":
        break   # 'break' interrompe o loop imediatamente

    # Adiciona o nome na lista
    convidados.append(nome)
    print(f"✅ {nome} adicionado! Total: {len(convidados)}\n")

# Mostra a lista final
print("\n🎉 Lista final de convidados:")
for convidado in convidados:
    print(f"  - {convidado}")

=== Cadastro de convidados ===
(digite 'sair' para parar)

Nome do convidado: Maria
✅ Maria adicionado! Total: 1

Nome do convidado: José
✅ José adicionado! Total: 2

Nome do convidado: Oscar
✅ Oscar adicionado! Total: 3

Nome do convidado: SAIR

🎉 Lista final de convidados:
  - Maria
  - José
  - Oscar


### ✏️ Exercício 1.2 — Adicionando elementos

Crie um programa de **cadastro de tarefas (to-do list)**:

1. Crie uma lista vazia chamada `tarefas`.
2. Use um `while` que pede ao usuário o nome de uma tarefa.
3. Enquanto o usuário não digitar `"fim"`, adicione cada tarefa na lista com `.append()`.
4. Ao final, mostre todas as tarefas cadastradas usando um `for` e `enumerate()` para numerá-las assim:

```
1 - Estudar Python
2 - Fazer compras
3 - Ligar para cliente
```

**Dica:** `enumerate(lista, start=1)` começa a contagem em 1 (em vez de 0).

In [ ]:
# 👉 Escreva seu código aqui:

# 1. Criar lista vazia
tarefas = []

enumerate(tarefas, start=1)
print("=== Cadastro de tarefas ===")
print("(digite 'fim' para parar)\n")

while True:   # Loop que só para com 'break'
    nome = input("Nome da tarefa: ")

    # Se o usuário digitou 'fim', sai do loop
    if nome.lower() == "fim":
        break   # 'break' interrompe o loop imediatamente

    tarefas.append(nome)
    print(f"✅ {nome} adicionado! Total: {len(tarefas)}\n")

# 3. Mostrar as tarefas numeradas usando for + enumerate
print("\n🎉 Lista final de tarefas:")
# Adicionando a verificação para listas vazias antes de iterar
if not tarefas:
    print("  Nenhuma tarefa cadastrada.")
else:
    for i, tarefa in enumerate(tarefas, start=1):
        print(f"  {i} - {tarefa}")

=== Cadastro de tarefas ===
(digite 'fim' para parar)

Nome da tarefa: Estudar Python
✅ Estudar Python adicionado! Total: 1

Nome da tarefa: Fazer Compras
✅ Fazer Compras adicionado! Total: 2

Nome da tarefa: Ligar para Clientes
✅ Ligar para Clientes adicionado! Total: 3

Nome da tarefa: fim

🎉 Lista final de tarefas:
  1 - Estudar Python
  2 - Fazer Compras
  3 - Ligar para Clientes


## 📌 1.3. Removendo elementos

Três formas principais de remover itens:

| Método | O que faz |
|--------|-----------|
| `.remove(item)` | Remove a **PRIMEIRA ocorrência** do item pelo **valor** |
| `.pop(índice)` | Remove pelo **ÍNDICE** e **retorna** o item removido (se não passar índice, remove o último) |
| `del lista[índice]` | Remove pelo índice (sem retornar nada) |
| `.clear()` | Apaga **todos** os itens (lista fica vazia) |

In [ ]:
# ============================================================
# EXEMPLO 1 - .remove() — remove pelo VALOR
# ============================================================

# Lista de frutas
frutas = ["maçã", "banana", "uva", "banana", "manga"]
print("Antes:", frutas)

# .remove() remove a PRIMEIRA ocorrência do valor indicado
frutas.remove("banana")   # Remove só a primeira 'banana'
print("Após remover 'banana':", frutas)

# CUIDADO: se o item NÃO existir na lista, dá erro!
# Para evitar, verificamos antes com 'in'
if "kiwi" in frutas:
    frutas.remove("kiwi")
else:
    print("Kiwi não está na lista, nada foi removido.")

Antes: ['maçã', 'banana', 'uva', 'banana', 'manga']
Após remover 'banana': ['maçã', 'uva', 'banana', 'manga']
Kiwi não está na lista, nada foi removido.


In [ ]:
# ============================================================
# EXEMPLO 2 - .pop() — remove pelo ÍNDICE e RETORNA o item
# ============================================================

# Lista de pessoas na fila
fila = ["Ana", "Bruno", "Carlos", "Diana"]
print("Fila inicial:", fila)

# .pop(0) remove o PRIMEIRO item e retorna ele
# Isso é útil quando queremos SABER quem saiu
atendido = fila.pop(0)
print(f"\nAgora atendendo: {atendido}")
print("Fila atualizada:", fila)

# .pop() SEM índice remove o ÚLTIMO item
ultimo = fila.pop()
print(f"\nÚltimo saiu: {ultimo}")
print("Fila final:", fila)

Fila inicial: ['Ana', 'Bruno', 'Carlos', 'Diana']

Agora atendendo: Ana
Fila atualizada: ['Bruno', 'Carlos', 'Diana']

Último saiu: Diana
Fila final: ['Bruno', 'Carlos']


In [ ]:
# ============================================================
# EXEMPLO 3 - del — remove pelo índice (sem retornar)
# ============================================================

# del é um comando do Python (não é método!)
# Funciona igual ao .pop() mas NÃO retorna o item removido

numeros = [10, 20, 30, 40, 50]
print("Antes:", numeros)

# Remove o item do índice 2 (o número 30)
del numeros[2]
print("Após del numeros[2]:", numeros)

# Também podemos deletar um 'pedaço' usando fatiamento
letras = ["a", "b", "c", "d", "e", "f"]
del letras[1:4]   # Remove do índice 1 até 3 (o 4 não inclui)
print("Após del letras[1:4]:", letras)   # ['a', 'e', 'f']

Antes: [10, 20, 30, 40, 50]
Após del numeros[2]: [10, 20, 40, 50]
Após del letras[1:4]: ['a', 'e', 'f']


In [ ]:
# ============================================================
# EXEMPLO 4 - .clear() + removendo com base em condição
# ============================================================

# .clear() apaga TODOS os itens, deixando a lista vazia
carrinho = ["camiseta", "tênis", "boné"]
print("Antes:", carrinho)

carrinho.clear()
print("Após .clear():", carrinho)   # []

print("\n---\n")

# Caso mais comum: remover itens que atendem a uma CONDIÇÃO
# Exemplo: remover notas abaixo de 6.0
notas = [7.5, 4.0, 9.2, 6.8, 3.5, 8.0, 5.5]
print("Notas originais:", notas)

# Criamos uma NOVA lista só com as notas aprovadas
# (essa é a forma mais segura de filtrar)
notas_aprovadas = []
for nota in notas:
    if nota >= 6.0:
        notas_aprovadas.append(nota)

print("Notas aprovadas:", notas_aprovadas)

Antes: ['camiseta', 'tênis', 'boné']
Após .clear(): []

---

Notas originais: [7.5, 4.0, 9.2, 6.8, 3.5, 8.0, 5.5]
Notas aprovadas: [7.5, 9.2, 6.8, 8.0]


### ✏️ Exercício 1.3 — Removendo elementos

Você é um **segurança de balada** 🕺 e tem uma lista de pessoas esperando para entrar:

```python
fila = ["Ana", "Bruno", "Carlos", "Diana", "Eduardo", "Fernanda", "Gabriel"]
```

Faça o seguinte:

1. Atenda a **primeira pessoa** da fila usando `.pop(0)` e mostre: `"Entrou: NOME"`.
2. Faça isso mais **2 vezes** (total de 3 pessoas atendidas).
3. **Carlos** desistiu — remova ele pelo nome usando `.remove()` (verifique se ele ainda está na fila!).
4. Use `.pop()` **sem índice** para atender a última pessoa da fila e mostre o nome.
5. No final, mostre **quem ainda está esperando** e **quantos sobraram**.

In [ ]:
# 👉 Escreva seu código aqui:

fila = ["Ana", "Bruno", "Carlos", "Diana", "Eduardo", "Fernanda", "Gabriel"]

# 1. Atender a primeira pessoa
print(f"Entrou: {fila.pop(0)}")

# 2. Atender mais 2 pessoas
print(f"Entrou: {fila.pop(0)}")
print(f"Entrou: {fila.pop(0)}")

# 3. Carlos desistiu — remova ele pelo nome usando .remove() (verifique se ele ainda está na fila!).
if "Carlos" in fila:
    fila.remove("Carlos")
    print("Carlos desistiu e foi removido da fila.")
else:
    print("Carlos já não estava mais na fila ou já havia sido atendido.")

# 4. Use .pop() sem índice para atender a última pessoa da fila e mostre o nome.
print(f"Entrou (último da fila): {fila.pop()}")

# 5. No final, mostre quem ainda está esperando e quantos sobraram.
print(f"\nQuem ainda está esperando: {fila}")
print(f"Total de pessoas esperando: {len(fila)}")

Entrou: Ana
Entrou: Bruno
Entrou: Carlos
Carlos já não estava mais na fila ou já havia sido atendido.
Entrou (último da fila): Gabriel

Quem ainda está esperando: ['Diana', 'Eduardo', 'Fernanda']
Total de pessoas esperando: 3


## 📌 1.4. Buscando e contando

Métodos úteis para **pesquisar** informações numa lista:

| Operador/Método | O que faz |
|--------|-----------|
| `item in lista` | Retorna `True` se o item existir na lista |
| `.index(item)` | Retorna o **índice** da primeira ocorrência |
| `.count(item)` | Conta **quantas vezes** o item aparece |
| `len(lista)` | Retorna o **tamanho** da lista |
| `max(lista)` / `min(lista)` | Maior / menor valor |
| `sum(lista)` | Soma todos os valores (só funciona com números) |

In [ ]:
# ============================================================
# EXEMPLO 1 - Operador 'in' — verificar se item existe
# ============================================================

# 'in' retorna True se o item estiver na lista, False se não
# Essa é a forma mais simples e rápida de verificar existência

frutas = ["maçã", "banana", "uva", "laranja"]

# Verificação simples
print("maçã" in frutas)       # True
print("kiwi" in frutas)       # False

# Uso prático dentro de um if
fruta_buscada = input("Digite uma fruta para buscar: ")

if fruta_buscada in frutas:
    print(f"✅ Temos {fruta_buscada}!")
else:
    print(f"❌ Não temos {fruta_buscada} em estoque.")

# 'not in' verifica se o item NÃO está na lista
if "kiwi" not in frutas:
    print("Precisamos comprar kiwi!")

True
False
Digite uma fruta para buscar: kiwi
❌ Não temos kiwi em estoque.
Precisamos comprar kiwi!


In [ ]:
# ============================================================
# EXEMPLO 2 - .index() — descobrir a POSIÇÃO de um item
# ============================================================

# .index() retorna o índice da PRIMEIRA ocorrência
# CUIDADO: se o item não existir, dá erro!

nomes = ["Ana", "Bruno", "Carlos", "Diana", "Bruno"]

# Descobrir em qual posição Carlos está
posicao = nomes.index("Carlos")
print(f"Carlos está na posição {posicao}")   # 2

# 'Bruno' aparece duas vezes, mas .index() retorna SÓ A PRIMEIRA
posicao_bruno = nomes.index("Bruno")
print(f"Primeiro Bruno está na posição {posicao_bruno}")   # 1

# Para evitar erro, SEMPRE verifique com 'in' antes de usar .index()
nome_buscado = "Eduardo"
if nome_buscado in nomes:
    pos = nomes.index(nome_buscado)
    print(f"{nome_buscado} está na posição {pos}")
else:
    print(f"{nome_buscado} não está na lista.")

Carlos está na posição 2
Primeiro Bruno está na posição 1
Eduardo não está na lista.


In [ ]:
# ============================================================
# EXEMPLO 3 - .count() — contar ocorrências
# ============================================================

# .count() retorna QUANTAS VEZES o item aparece na lista
# Se o item não existir, retorna 0 (nunca dá erro)

votos = ["Ana", "Bruno", "Ana", "Carlos", "Ana", "Bruno", "Diana", "Ana"]

# Contando votos de cada candidato
print(f"Ana teve {votos.count('Ana')} votos")         # 4
print(f"Bruno teve {votos.count('Bruno')} votos")     # 2
print(f"Carlos teve {votos.count('Carlos')} votos")   # 1
print(f"Diana teve {votos.count('Diana')} votos")     # 1
print(f"Eduardo teve {votos.count('Eduardo')} votos") # 0 (não existe)

# Total de votos
print(f"\nTotal de votos: {len(votos)}")

Ana teve 4 votos
Bruno teve 2 votos
Carlos teve 1 votos
Diana teve 1 votos
Eduardo teve 0 votos

Total de votos: 8


In [ ]:
# ============================================================
# EXEMPLO 4 - Funções max(), min(), sum() com listas numéricas
# ============================================================

# Python tem funções prontas que trabalham com listas numéricas
notas = [7.5, 4.0, 9.2, 6.8, 3.5, 8.0, 5.5, 10.0]

# max() → maior valor
maior = max(notas)
print(f"Maior nota: {maior}")

# min() → menor valor
menor = min(notas)
print(f"Menor nota: {menor}")

# sum() → soma de todos os valores
total = sum(notas)
print(f"Soma das notas: {total}")

# Média = soma / quantidade
media = sum(notas) / len(notas)
print(f"Média: {media:.2f}")

# Amplitude = maior - menor
amplitude = max(notas) - min(notas)
print(f"Amplitude: {amplitude}")

Maior nota: 10.0
Menor nota: 3.5
Soma das notas: 54.5
Média: 6.81
Amplitude: 6.5


### ✏️ Exercício 1.4 — Buscando e contando

Você tem a lista de notas de uma prova:

```python
notas = [7.5, 4.0, 9.2, 6.8, 10.0, 3.5, 8.0, 5.5, 10.0, 6.0, 9.2]
```

Faça:

1. Mostre a **maior** e a **menor** nota.
2. Mostre a **média** da turma (com 2 casas decimais).
3. Conte quantos alunos tiraram nota **10.0**.
4. Verifique se alguém tirou nota **zero** usando `in`.
5. Descubra em que **posição** está a primeira nota `9.2`.
6. Mostre quantos alunos **passaram** (nota ≥ 6.0) usando `for` + `if`.

In [ ]:
# 👉 Escreva seu código aqui:

notas = [7.5, 4.0, 9.2, 6.8, 10.0, 3.5, 8.0, 5.5, 10.0, 6.0, 9.2]

# 1. Maior e menor nota
print(f"Maior nota: {max(notas)}")
print(f"Menor nota: {min(notas)}")

# 2. Média (2 casas decimais)
print(f"Média: {sum(notas) / len(notas):.2f}")

# 3. Quantos tiraram 10
print(f"Alunos com 10.0: {notas.count(10.0)}")

# 4. Alguém tirou zero?
print(f"Alguém tirou zero? {'Sim' if 0.0 in notas else 'Não'}")

# 5. Posição da primeira nota 9.2
print(f"Posição da primeira nota 9.2: {notas.index(9.2)}")

# 6. Quantos passaram (for + if)
print("Alunos que passaram:")
for nota in notas:
    if nota >= 6.0:
        print(f"  - {nota}")


Maior nota: 10.0
Menor nota: 3.5
Média: 7.25
Alunos com 10.0: 2
Alguém tirou zero? Não
Posição da primeira nota 9.2: 2
Alunos que passaram:
  - 7.5
  - 9.2
  - 6.8
  - 10.0
  - 8.0
  - 10.0
  - 6.0
  - 9.2


## 📌 1.5. Ordenando e invertendo

Métodos para **reorganizar** a lista:

| Método | O que faz |
|--------|-----------|
| `.sort()` | Ordena a lista **em ordem crescente** (modifica a lista original) |
| `.sort(reverse=True)` | Ordena em ordem **decrescente** |
| `sorted(lista)` | Retorna uma **nova lista ordenada** (sem alterar a original) |
| `.reverse()` | **Inverte** a ordem dos itens |

**Diferença importante:** `.sort()` **MODIFICA** a lista original. `sorted()` retorna uma **NOVA** lista.

In [ ]:
# ============================================================
# EXEMPLO 1 - .sort() — ordena CRESCENTE
# ============================================================

# Com números
numeros = [5, 2, 9, 1, 7, 3, 8]
print("Antes:", numeros)

# .sort() MODIFICA a lista original (não retorna nada!)
numeros.sort()
print("Depois:", numeros)   # [1, 2, 3, 5, 7, 8, 9]

# Com textos (ordena em ordem alfabética)
nomes = ["Carlos", "Ana", "Diana", "Bruno"]
nomes.sort()
print("Nomes ordenados:", nomes)   # ['Ana', 'Bruno', 'Carlos', 'Diana']

Antes: [5, 2, 9, 1, 7, 3, 8]
Depois: [1, 2, 3, 5, 7, 8, 9]
Nomes ordenados: ['Ana', 'Bruno', 'Carlos', 'Diana']


In [ ]:
# ============================================================
# EXEMPLO 2 - .sort(reverse=True) — ordena DECRESCENTE
# ============================================================

# Passando reverse=True, ordena do maior pro menor
notas = [7.5, 4.0, 9.2, 6.8, 10.0, 3.5]

notas.sort(reverse=True)
print("Notas do maior para o menor:", notas)
# Resultado: [10.0, 9.2, 7.5, 6.8, 4.0, 3.5]

# Útil para rankings!
print("\n🏆 Top 3 notas:")
for i in range(3):
    print(f"{i + 1}º lugar: {notas[i]}")

Notas do maior para o menor: [10.0, 9.2, 7.5, 6.8, 4.0, 3.5]

🏆 Top 3 notas:
1º lugar: 10.0
2º lugar: 9.2
3º lugar: 7.5


In [ ]:
# ============================================================
# EXEMPLO 3 - sorted() — retorna NOVA lista ordenada
# ============================================================

# sorted() NÃO modifica a lista original, cria uma nova
# Útil quando queremos preservar a ordem original

precos = [29.90, 15.50, 99.00, 8.75, 45.00]

# Cria uma nova lista ordenada sem mexer na original
precos_ordenados = sorted(precos)

print("Original:", precos)                    # Mantém [29.90, 15.50, ...]
print("Ordenada:", precos_ordenados)          # Nova lista ordenada

# Também aceita reverse=True
precos_desc = sorted(precos, reverse=True)
print("Decrescente:", precos_desc)

Original: [29.9, 15.5, 99.0, 8.75, 45.0]
Ordenada: [8.75, 15.5, 29.9, 45.0, 99.0]
Decrescente: [99.0, 45.0, 29.9, 15.5, 8.75]


In [ ]:
# ============================================================
# EXEMPLO 4 - .reverse() — inverte a ordem
# ============================================================

# .reverse() inverte a lista (último vira primeiro, primeiro vira último)
# NÃO ordena! Apenas inverte a ordem ATUAL

letras = ["a", "b", "c", "d", "e"]
letras.reverse()
print("Invertida:", letras)   # ['e', 'd', 'c', 'b', 'a']

# CUIDADO: reverse é diferente de sort(reverse=True)!
# .reverse() só INVERTE a ordem atual
# .sort(reverse=True) ORDENA do maior pro menor

numeros = [3, 1, 4, 1, 5, 9, 2, 6]

# Versão A: só invertendo
copia_a = numeros.copy()   # .copy() cria uma cópia da lista
copia_a.reverse()
print(".reverse():", copia_a)   # [6, 2, 9, 5, 1, 4, 1, 3] — só inverteu

# Versão B: ordenando decrescente
copia_b = numeros.copy()
copia_b.sort(reverse=True)
print(".sort(reverse=True):", copia_b)   # [9, 6, 5, 4, 3, 2, 1, 1] — ordenou

Invertida: ['e', 'd', 'c', 'b', 'a']
.reverse(): [6, 2, 9, 5, 1, 4, 1, 3]
.sort(reverse=True): [9, 6, 5, 4, 3, 2, 1, 1]


### ✏️ Exercício 1.5 — Ordenando e invertendo

Você tem as idades dos participantes de um evento:

```python
idades = [25, 18, 42, 30, 27, 55, 19, 36, 22, 48]
```

Faça:

1. Crie uma **nova lista ordenada** (do menor para o maior) usando `sorted()`, **SEM alterar** a original.
2. Mostre a lista **original** e a **ordenada** para comprovar que a original não mudou.
3. Ordene a lista **original** em ordem **decrescente** usando `.sort(reverse=True)`.
4. Mostre as **3 pessoas mais velhas** (use fatiamento).
5. **Inverta** a lista com `.reverse()` e mostre o resultado.

In [ ]:
# 👉 Escreva seu código aqui:

idades = [25, 18, 42, 30, 27, 55, 19, 36, 22, 48]

# 1. Nova lista ordenada com sorted()
print("Original:", idades)
print("Ordenada:", sorted(idades))

# 2. Mostrar original e ordenada
print("\nOriginal:", idades)
print("Ordenada:", sorted(idades))

# 3. Ordenar a original em ordem decrescente
print("\nOriginal:", idades)
idades.sort(reverse=True)
print("Ordenada:", idades)

# 4. Mostrar as 3 mais velhas (fatiamento)
print("\n3 pessoas mais velhas:", idades[:3])

# 5. Inverter com .reverse()
print("\nOriginal:", idades)
idades.reverse()
print("Invertida:", idades)


Original: [25, 18, 42, 30, 27, 55, 19, 36, 22, 48]
Ordenada: [18, 19, 22, 25, 27, 30, 36, 42, 48, 55]

Original: [25, 18, 42, 30, 27, 55, 19, 36, 22, 48]
Ordenada: [18, 19, 22, 25, 27, 30, 36, 42, 48, 55]

Original: [25, 18, 42, 30, 27, 55, 19, 36, 22, 48]
Ordenada: [55, 48, 42, 36, 30, 27, 25, 22, 19, 18]

3 pessoas mais velhas: [55, 48, 42]

Original: [55, 48, 42, 36, 30, 27, 25, 22, 19, 18]
Invertida: [18, 19, 22, 25, 27, 30, 36, 42, 48, 55]


---
# 📚 PARTE 2 — DICIONÁRIOS

Enquanto listas guardam itens em **posições numeradas** (0, 1, 2...), dicionários guardam itens em pares **chave → valor**.

**Pense em um dicionário comum:** você procura uma palavra (**chave**) e encontra o significado (**valor**). Python funciona igual!

**Sintaxe:**
```python
pessoa = {
    "nome": "Ana",       # chave: "nome"  → valor: "Ana"
    "idade": 25,          # chave: "idade" → valor: 25
    "cidade": "Recife"    # chave: "cidade" → valor: "Recife"
}
```

**Quando usar lista vs dicionário?**
- **Lista** → quando os itens formam uma **sequência** (nomes de alunos, temperaturas, etc.)
- **Dicionário** → quando você quer **associar** informações relacionadas (nome → idade, produto → preço)

## 📌 2.1. Criando e acessando dicionários

**Características:**
- Criados com **chaves** `{ }`
- Cada item é um par `chave: valor`
- Pares separados por **vírgula**
- Acessamos o valor usando `dicionario[chave]`
- Chaves são **únicas** (não pode ter duas chaves iguais)

In [ ]:
# ============================================================
# EXEMPLO 1 - Criando e acessando um dicionário simples
# ============================================================

# Criando um dicionário com dados de uma pessoa
# As chaves são os 'rótulos' e os valores são as 'informações'
pessoa = {
    "nome": "Maria Silva",
    "idade": 30,
    "profissao": "Engenheira",
    "salario": 5500.00
}

# Para acessar um valor, usamos a CHAVE (não o índice!)
# Reparem que usamos colchetes [] mas com uma chave de texto dentro
print(pessoa["nome"])      # Maria Silva
print(pessoa["idade"])     # 30
print(pessoa["profissao"]) # Engenheira

# Bom uso com f-string:
print(f"\n{pessoa['nome']} tem {pessoa['idade']} anos e é {pessoa['profissao']}.")

Maria Silva
30
Engenheira

Maria Silva tem 30 anos e é Engenheira.


In [ ]:
# ============================================================
# EXEMPLO 2 - .get() — forma SEGURA de acessar valores
# ============================================================

# Se a chave NÃO existir, acessar com [] dá erro (KeyError)!
# O método .get() é mais seguro: se não existir, retorna None (ou um valor padrão)

produto = {
    "nome": "Notebook",
    "preco": 3500.00,
    "estoque": 15
}

# Forma normal (dá erro se a chave não existir)
print(produto["nome"])   # OK, existe

# Forma segura com .get()
cor = produto.get("cor")           # Retorna None (não dá erro)
print(f"Cor: {cor}")

# Podemos definir um valor padrão caso a chave não exista
cor = produto.get("cor", "não informada")
print(f"Cor: {cor}")   # não informada

# Útil em condicionais
if produto.get("desconto"):
    print("Tem desconto!")
else:
    print("Sem desconto.")

Notebook
Cor: None
Cor: não informada
Sem desconto.


In [ ]:
# ============================================================
# EXEMPLO 3 - Adicionando e modificando pares
# ============================================================

# Dicionários são MUTÁVEIS: podemos adicionar, modificar e remover

aluno = {
    "nome": "João",
    "idade": 20
}
print("Inicial:", aluno)

# ADICIONAR nova chave: basta atribuir um valor a uma chave nova
aluno["curso"] = "Engenharia"
aluno["nota"] = 8.5
print("Com mais dados:", aluno)

# MODIFICAR valor existente: mesma sintaxe, mas com chave que já existe
aluno["idade"] = 21   # Fez aniversário
aluno["nota"] = 9.0   # Melhorou a nota
print("Após modificações:", aluno)

# REMOVER uma chave: usamos 'del' ou .pop()
del aluno["nota"]
print("Sem a nota:", aluno)

Inicial: {'nome': 'João', 'idade': 20}
Com mais dados: {'nome': 'João', 'idade': 20, 'curso': 'Engenharia', 'nota': 8.5}
Após modificações: {'nome': 'João', 'idade': 21, 'curso': 'Engenharia', 'nota': 9.0}
Sem a nota: {'nome': 'João', 'idade': 21, 'curso': 'Engenharia'}


In [ ]:
# ============================================================
# EXEMPLO 4 - Dicionário DENTRO de lista (muito comum!)
# ============================================================

# É super comum termos uma LISTA de DICIONÁRIOS
# Exemplo: uma lista de alunos, onde cada aluno é um dicionário

alunos = [
    {"nome": "Ana", "idade": 20, "nota": 8.5},
    {"nome": "Bruno", "idade": 22, "nota": 7.0},
    {"nome": "Carlos", "idade": 19, "nota": 9.2},
]

# Acessando um aluno específico pelo índice da lista
# e depois pegando um valor específico pelo nome da chave
print(alunos[0]["nome"])   # Ana
print(alunos[2]["nota"])   # 9.2

# Percorrendo a lista com for
print("\n📋 Lista de alunos:")
for aluno in alunos:
    # Dentro do for, 'aluno' é um dicionário
    print(f"- {aluno['nome']} ({aluno['idade']} anos) → Nota: {aluno['nota']}")

Ana
9.2

📋 Lista de alunos:
- Ana (20 anos) → Nota: 8.5
- Bruno (22 anos) → Nota: 7.0
- Carlos (19 anos) → Nota: 9.2


### ✏️ Exercício 2.1 — Criando e acessando dicionários

Crie um dicionário `carro` representando o seu carro dos sonhos com as chaves:
- `"marca"` (ex: `"Toyota"`)
- `"modelo"` (ex: `"Corolla"`)
- `"ano"` (ex: `2024`)
- `"cor"` (ex: `"prata"`)
- `"preco"` (ex: `150000.00`)

Depois faça:

1. Mostre cada informação em uma linha, usando f-string.
2. Adicione uma nova chave `"combustivel"` com o valor `"flex"`.
3. Modifique o `"preco"` dando um desconto de R$ 10.000 (use `-=`).
4. Use `.get()` para tentar acessar a chave `"quilometragem"` com valor padrão `"0 km"`.
5. Mostre o dicionário completo no final.

In [ ]:
# 👉 Escreva seu código aqui:

# Criar o dicionário carro
carro = {
    "marca": "Toyota",
    "modelo": "Corolla",
    "ano": 2024,
    "cor": "prata",
    "preco": 150000.00
}

# 1. Mostrar cada informação
print(f"Marca: {carro['marca']}")
print(f"Modelo: {carro['modelo']}")
print(f"Ano: {carro['ano']}")

# 2. Adicionar combustivel
print("\nCombustivel: flex")
carro["combustivel"] = "flex"

# 3. Aplicar desconto no preço
print("\nDesconto de R$ 10.000")
carro["preco"] -= 10000

# 4. .get() com valor padrão para quilometragem
quilometragem_info = carro.get("quilometragem", "0 km")
print("\nQuilometragem:", quilometragem_info)

# 5. Mostrar dicionário completo
print("\nCarro atualizado:", carro)

Marca: Toyota
Modelo: Corolla
Ano: 2024

Combustivel: flex

Desconto de R$ 10.000

Quilometragem: 0 km

Carro atualizado: {'marca': 'Toyota', 'modelo': 'Corolla', 'ano': 2024, 'cor': 'prata', 'preco': 140000.0, 'combustivel': 'flex'}


## 📌 2.2. Percorrendo dicionários com for

Três formas de percorrer um dicionário com `for`:

| Método | O que retorna |
|--------|---------------|
| `dicionario` (sozinho) | Só as **chaves** |
| `.keys()` | Só as **chaves** (mesma coisa) |
| `.values()` | Só os **valores** |
| `.items()` | **Pares** chave-valor (mais usado!) |

In [ ]:
# ============================================================
# EXEMPLO 1 - Percorrendo só as CHAVES
# ============================================================

precos = {
    "pão": 0.50,
    "leite": 6.00,
    "ovo": 0.80,
    "café": 15.00
}

# Forma 1: for direto no dicionário (pega só as chaves)
print("Produtos disponíveis:")
for produto in precos:
    print(f"- {produto}")

# Forma 2: usando .keys() (é a mesma coisa, mas mais explícito)
print("\nMesma coisa com .keys():")
for chave in precos.keys():
    print(f"- {chave}")

In [ ]:
# ============================================================
# EXEMPLO 2 - Percorrendo só os VALORES
# ============================================================

# .values() nos dá só os valores (sem as chaves)
# Útil quando só queremos os dados, sem os 'rótulos'

precos = {
    "pão": 0.50,
    "leite": 6.00,
    "ovo": 0.80,
    "café": 15.00
}

# Calculando o total gasto (só precisamos dos VALORES)
total = 0
for preco in precos.values():
    total += preco

print(f"Total: R$ {total:.2f}")

# Dá pra usar sum() direto com .values()!
total_rapido = sum(precos.values())
print(f"Total (com sum): R$ {total_rapido:.2f}")

# Mais alguns usos
print(f"Maior preço: R$ {max(precos.values()):.2f}")
print(f"Menor preço: R$ {min(precos.values()):.2f}")

In [ ]:
# ============================================================
# EXEMPLO 3 - .items() — chave E valor ao mesmo tempo
# ============================================================

# .items() é o MAIS USADO: dá chave E valor de uma vez!
# Usamos DUAS variáveis no for: uma para cada

precos = {
    "pão": 0.50,
    "leite": 6.00,
    "ovo": 0.80,
    "café": 15.00
}

print("📋 Cardápio:")
# A cada volta: produto = chave atual, preco = valor atual
for produto, preco in precos.items():
    print(f"  {produto:.<15} R$ {preco:>6.2f}")

# O ':.<15' alinha à esquerda com '.' de preenchimento em 15 chars
# O ':>6.2f' alinha à direita com 6 posições e 2 casas decimais
# Isso deixa a saída BEM bonitinha, tipo um cardápio real!

In [ ]:
# ============================================================
# EXEMPLO 4 - Aplicando condições ao percorrer
# ============================================================

# Dicionário com pontuação dos alunos em um jogo
pontuacoes = {
    "Ana": 850,
    "Bruno": 420,
    "Carlos": 1200,
    "Diana": 680,
    "Eduardo": 990,
    "Fernanda": 350
}

# Filtrando quem tem mais de 700 pontos (classificado)
print("🏆 Classificados (>700 pontos):")
classificados = 0
for nome, pontos in pontuacoes.items():
    if pontos > 700:
        print(f"  ✅ {nome}: {pontos} pts")
        classificados += 1

print(f"\nTotal classificados: {classificados}")

# Encontrando o(a) campeão(ã)
campeao = ""
maior_pontuacao = 0
for nome, pontos in pontuacoes.items():
    if pontos > maior_pontuacao:
        maior_pontuacao = pontos
        campeao = nome

print(f"\n👑 Campeão(ã): {campeao} com {maior_pontuacao} pontos!")

### ✏️ Exercício 2.2 — Percorrendo dicionários

Você tem o estoque de uma loja:

```python
estoque = {
    "camiseta": 50,
    "calça": 20,
    "tênis": 15,
    "boné": 35,
    "meia": 80,
    "jaqueta": 8
}
```

Faça:

1. Percorra com `.items()` e mostre cada produto e quantidade: `"Produto: X | Quantidade: Y"`.
2. Mostre o **total geral de itens** usando `sum(estoque.values())`.
3. Liste os produtos com estoque **baixo** (menos de 20 unidades).
4. Encontre o produto com **MAIOR estoque** (nome e quantidade).

In [ ]:
# 👉 Escreva seu código aqui:

estoque = {
    "camiseta": 50,
    "calça": 20,
    "tênis": 15,
    "boné": 35,
    "meia": 80,
    "jaqueta": 8
}

# 1. Percorrer com .items() e mostrar cada produto
print("Estoque:")
for produto, quantidade in estoque.items():
    print(f"  Produto: {produto} | Quantidade: {quantidade}")

# 2. Total geral de itens
print(f"\nTotal geral de itens: {sum(estoque.values())}")

# 3. Produtos com estoque baixo (< 20)
print("\nProdutos com estoque baixo:")
for produto, quantidade in estoque.items():
    if quantidade < 20:
        print(f"  {produto}: {quantidade} unidades")

# 4. Produto com maior estoque
print("\nProduto com maior estoque:")
produto_maior_estoque = max(estoque, key=estoque.get)
print(f"  {produto_maior_estoque}: {estoque[produto_maior_estoque]} unidades")


Estoque:
  Produto: camiseta | Quantidade: 50
  Produto: calça | Quantidade: 20
  Produto: tênis | Quantidade: 15
  Produto: boné | Quantidade: 35
  Produto: meia | Quantidade: 80
  Produto: jaqueta | Quantidade: 8

Total geral de itens: 208

Produtos com estoque baixo:
  tênis: 15 unidades
  jaqueta: 8 unidades

Produto com maior estoque:
  meia: 80 unidades


## 📌 2.3. Métodos e operações com dicionários

Métodos úteis:

| Método | O que faz |
|--------|-----------|
| `chave in dicionario` | Verifica se a **chave** existe |
| `len(dicionario)` | Quantidade de pares |
| `.pop(chave)` | Remove e retorna o valor |
| `.update(outro_dict)` | Mescla dois dicionários |
| `.clear()` | Esvazia o dicionário |

In [ ]:
# ============================================================
# EXEMPLO 1 - Verificando existência com 'in'
# ============================================================

# 'in' verifica se uma CHAVE existe no dicionário (não o valor!)

usuario = {
    "nome": "Carla",
    "email": "carla@email.com",
    "idade": 28
}

# Verifica se uma chave existe
if "email" in usuario:
    print(f"Email cadastrado: {usuario['email']}")
else:
    print("Email não cadastrado.")

# 'not in' verifica se NÃO existe
if "telefone" not in usuario:
    print("Telefone não cadastrado. Vamos adicionar!")
    usuario["telefone"] = "99999-9999"

print("\nUsuário atualizado:", usuario)

In [ ]:
# ============================================================
# EXEMPLO 2 - .pop() para remover e salvar o valor
# ============================================================

# .pop(chave) remove o par E retorna o valor removido
# Similar ao .pop() das listas, mas usando CHAVE em vez de índice

carrinho = {
    "camiseta": 49.90,
    "calça": 89.90,
    "tênis": 199.90
}
print("Carrinho inicial:", carrinho)

# O cliente desistiu da calça
preco_removido = carrinho.pop("calça")
print(f"\nItem removido custava: R$ {preco_removido:.2f}")
print("Carrinho agora:", carrinho)

# .pop() com valor padrão (se a chave não existir, não dá erro)
preco_boné = carrinho.pop("boné", 0.0)
print(f"Boné retornou: {preco_boné} (não estava no carrinho)")

In [ ]:
# ============================================================
# EXEMPLO 3 - .update() para mesclar dicionários
# ============================================================

# .update() junta outro dicionário no atual
# Se a chave já existir, o valor é ATUALIZADO
# Se não existir, a chave é ADICIONADA

pessoa = {
    "nome": "Pedro",
    "idade": 25,
    "cidade": "Porto Alegre"
}

novos_dados = {
    "idade": 26,              # Atualiza a idade existente
    "profissao": "Professor", # Adiciona nova chave
    "estado": "RS"            # Adiciona nova chave
}

pessoa.update(novos_dados)

print("Pessoa atualizada:")
for chave, valor in pessoa.items():
    print(f"  {chave}: {valor}")

In [ ]:
# ============================================================
# EXEMPLO 4 - Contando ocorrências com dicionário
# ============================================================

# Uma aplicação super comum: contar quantas vezes cada item aparece
# Por exemplo: contar letras em uma palavra

frase = "programacao"
contagem = {}   # Começamos com dicionário vazio

# Percorremos cada letra da frase
for letra in frase:
    if letra in contagem:
        # Se a letra já está no dicionário, incrementa
        contagem[letra] += 1
    else:
        # Se é a primeira vez, cria com valor 1
        contagem[letra] = 1

print(f"Contagem de letras em '{frase}':")
for letra, qtd in contagem.items():
    print(f"  '{letra}' → {qtd} vez(es)")

# Forma MAIS SIMPLES usando .get() com valor padrão 0
frase2 = "banana"
contagem2 = {}
for letra in frase2:
    # Se a letra não existe, .get() retorna 0, e somamos 1
    contagem2[letra] = contagem2.get(letra, 0) + 1

print(f"\nContagem de '{frase2}': {contagem2}")

### ✏️ Exercício 2.3 — Métodos de dicionários

Você tem os votos de uma eleição de turma:

```python
votos = ["Ana", "Bruno", "Ana", "Carlos", "Ana", "Bruno", "Diana", "Ana", "Carlos", "Bruno", "Ana", "Diana"]
```

Usando um **dicionário** para contar os votos, faça:

1. Crie um dicionário vazio chamado `apuracao`.
2. Percorra a lista `votos` com `for` e conte quantos votos cada candidato teve (use o truque do `.get(chave, 0) + 1`).
3. Mostre o resultado da apuração ordenado (dica: use `.items()` com `for`).
4. Descubra e mostre quem foi o **vencedor** e quantos votos teve.
5. Mostre o **total de votos** usando `sum(apuracao.values())`.

In [ ]:
# 👉 Escreva seu código aqui:

votos = ["Ana", "Bruno", "Ana", "Carlos", "Ana", "Bruno", "Diana", "Ana", "Carlos", "Bruno", "Ana", "Diana"]

# 1. Dicionário vazio
apuracao = {} # Corrigido: deve ser um dicionário, não uma lista

# 2. Contar votos com for
print("Apuração:")
for candidato in votos:
    apuracao[candidato] = apuracao.get(candidato, 0) + 1

# 3. Mostrar apuração
print("\nApuração:")
for candidato, num_votos in apuracao.items(): # Renomeado 'votos' para 'num_votos' para evitar conflito
    print(f"  {candidato}: {num_votos} votos")

# 4. Descobrir vencedor
vencedor = max(apuracao, key=apuracao.get)
print(f"\nVencedor: {vencedor} com {apuracao[vencedor]} votos")

# 5. Total de votos
print(f"\nTotal de votos: {sum(apuracao.values())}")

Apuração:

Apuração:
  Ana: 5 votos
  Bruno: 3 votos
  Carlos: 2 votos
  Diana: 2 votos

Vencedor: Ana com 5 votos

Total de votos: 12


## 📌 2.4. Dicionários aninhados (dicionário dentro de dicionário)

Um dicionário pode ter **outros dicionários** como valores. Isso é perfeito para representar dados complexos do mundo real!

```python
pessoa = {
    "nome": "Ana",
    "endereco": {                    # ← aqui tem um dicionário dentro
        "rua": "Av. Brasil",
        "numero": 123,
        "cidade": "Porto Alegre"
    }
}
```

In [ ]:
# ============================================================
# EXEMPLO 1 - Acessando dados em dicionário aninhado
# ============================================================

funcionario = {
    "nome": "Carla",
    "cargo": "Gerente",
    "endereco": {
        "rua": "Rua das Flores",
        "numero": 456,
        "cidade": "Recife",
        "cep": "50000-000"
    },
    "contato": {
        "email": "carla@empresa.com",
        "telefone": "81 99999-1234"
    }
}

# Para acessar dados aninhados, usamos colchetes em cadeia
# Primeiro acessamos o dicionário interno, depois a chave dentro dele
print(funcionario["nome"])                       # Carla
print(funcionario["endereco"]["cidade"])          # Recife
print(funcionario["contato"]["email"])            # carla@empresa.com

# Com f-string (muito mais legível!)
print(f"\n{funcionario['nome']} mora em {funcionario['endereco']['cidade']}")
print(f"Email: {funcionario['contato']['email']}")

In [ ]:
# ============================================================
# EXEMPLO 2 - Lista de dicionários aninhados
# ============================================================

# Exemplo REAL: banco de dados simples de produtos
produtos = [
    {
        "id": 1,
        "nome": "Notebook",
        "preco": 3500.00,
        "categoria": "Eletrônicos"
    },
    {
        "id": 2,
        "nome": "Camiseta",
        "preco": 49.90,
        "categoria": "Roupas"
    },
    {
        "id": 3,
        "nome": "Café",
        "preco": 15.00,
        "categoria": "Alimentos"
    },
]

# Percorrendo e formatando bonito
print("📋 CATÁLOGO DE PRODUTOS")
print("=" * 50)
for produto in produtos:
    print(f"#{produto['id']} - {produto['nome']}")
    print(f"   Categoria: {produto['categoria']}")
    print(f"   Preço: R$ {produto['preco']:.2f}")
    print("-" * 50)

In [ ]:
# ============================================================
# EXEMPLO 3 - Agenda de contatos aninhada
# ============================================================

# Dicionário onde cada chave é um nome e o valor é um dicionário
# com os dados daquela pessoa

agenda = {
    "Ana": {
        "telefone": "51 99111-1111",
        "email": "ana@email.com",
        "aniversario": "15/03"
    },
    "Bruno": {
        "telefone": "51 99222-2222",
        "email": "bruno@email.com",
        "aniversario": "22/07"
    },
    "Carlos": {
        "telefone": "51 99333-3333",
        "email": "carlos@email.com",
        "aniversario": "10/12"
    }
}

# Buscando dados de uma pessoa específica
nome_buscado = "Ana"
if nome_buscado in agenda:
    dados = agenda[nome_buscado]
    print(f"📞 Contato de {nome_buscado}:")
    print(f"  Telefone: {dados['telefone']}")
    print(f"  Email: {dados['email']}")
    print(f"  Aniversário: {dados['aniversario']}")

# Listando todos os aniversariantes
print("\n🎂 Aniversariantes:")
for nome, dados in agenda.items():
    print(f"  {nome}: {dados['aniversario']}")

In [ ]:
# ============================================================
# EXEMPLO 4 - Boletim escolar com dicionário aninhado
# ============================================================

# Cada aluno tem um dicionário com várias matérias e notas
boletim = {
    "Ana": {"matemática": 8.5, "português": 9.0, "história": 7.5},
    "Bruno": {"matemática": 6.0, "português": 5.5, "história": 7.0},
    "Carlos": {"matemática": 9.8, "português": 8.0, "história": 9.5},
}

# Calculando a média de cada aluno
print("📚 BOLETIM ESCOLAR")
print("=" * 50)
for aluno, notas in boletim.items():
    # notas é um dicionário. Usamos sum() com .values() e len() para calcular a média
    media = sum(notas.values()) / len(notas)
    situacao = "Aprovado ✅" if media >= 7.0 else "Recuperação ⚠️"

    print(f"\n👤 {aluno}")
    for materia, nota in notas.items():
        print(f"  {materia.capitalize():.<20} {nota}")
    print(f"  {'MÉDIA':.<20} {media:.2f} → {situacao}")

### ✏️ Exercício 2.4 — Dicionários aninhados

Crie uma **agenda telefônica** usando dicionário aninhado:

```python
agenda = {
    "Ana": {"telefone": "51 99111-1111", "cidade": "Porto Alegre"},
    "Bruno": {"telefone": "11 99222-2222", "cidade": "São Paulo"},
    "Carlos": {"telefone": "21 99333-3333", "cidade": "Rio de Janeiro"},
    "Diana": {"telefone": "51 99444-4444", "cidade": "Porto Alegre"}
}
```

Faça:

1. Percorra a agenda com `.items()` e mostre cada contato formatado: `"NOME — TELEFONE (CIDADE)"`.
2. Adicione um novo contato (`"Eduardo"` com telefone e cidade à sua escolha).
3. Usando `for` + `if`, mostre **apenas** os contatos que moram em `"Porto Alegre"`.
4. Peça ao usuário um nome com `input()` e mostre os dados dessa pessoa (use `in` para verificar se existe).

In [ ]:
# 👉 Escreva seu código aqui:

agenda = {
    "Ana": {"telefone": "51 99111-1111", "cidade": "Porto Alegre"},
    "Bruno": {"telefone": "11 99222-2222", "cidade": "São Paulo"},
    "Carlos": {"telefone": "21 99333-3333", "cidade": "Rio de Janeiro"},
    "Diana": {"telefone": "51 99444-4444", "cidade": "Porto Alegre"}
}

# 1. Mostrar cada contato formatado
print("Agenda Telefônica:")
for nome, dados in agenda.items():
    print(f"{nome} — {dados['telefone']} ({dados['cidade']})")

# 2. Adicionar Eduardo
print("\n---\n")
agenda["Eduardo"] = {"telefone": "51 91234-5678", "cidade": "Florianópolis"}
print("Agenda atualizada:")
for nome, dados in agenda.items():
    print(f"{nome} — {dados['telefone']} ({dados['cidade']})")

# 3. Mostrar só quem é de Porto Alegre
print("\n---\n")
print("Contatos de Porto Alegre:")
for nome, dados in agenda.items():
    if dados["cidade"] == "Porto Alegre":
        print(f"{nome} — {dados['telefone']}")

# 4. Buscar por nome via input()
input_nome = input("\nDigite o nome para buscar: ")
if input_nome in agenda:
    print(f"Informações de {input_nome}:")
    print(f"  Telefone: {agenda[input_nome]['telefone']}")


Agenda Telefônica:
Ana — 51 99111-1111 (Porto Alegre)
Bruno — 11 99222-2222 (São Paulo)
Carlos — 21 99333-3333 (Rio de Janeiro)
Diana — 51 99444-4444 (Porto Alegre)

---

Agenda atualizada:
Ana — 51 99111-1111 (Porto Alegre)
Bruno — 11 99222-2222 (São Paulo)
Carlos — 21 99333-3333 (Rio de Janeiro)
Diana — 51 99444-4444 (Porto Alegre)
Eduardo — 51 91234-5678 (Florianópolis)

---

Contatos de Porto Alegre:
Ana — 51 99111-1111
Diana — 51 99444-4444

Digite o nome para buscar: Diana
Informações de Diana:
  Telefone: 51 99444-4444


---
# 🔧 PARTE 3 — FUNÇÕES

**Função** é um **bloco de código reutilizável** que executa uma tarefa específica. Você **define uma vez** e pode **chamar quantas vezes quiser**.

**Por que usar funções?**
- ✅ **Evita repetição**: se você vai calcular a mesma coisa 10 vezes, basta criar uma função
- ✅ **Organiza o código**: separa o programa em pedaços menores e mais claros
- ✅ **Facilita manutenção**: se precisar mudar algo, muda em um lugar só
- ✅ **Reaproveita em outros projetos**

**Sintaxe básica:**
```python
def nome_da_funcao(parametros):
    # bloco de código
    return resultado   # opcional

# Chamando a função
resultado = nome_da_funcao(valores)
```

**Vocabulário:**
- **Definir** uma função → criar usando `def`
- **Chamar** uma função → usar/executar a função
- **Parâmetro** → variável que a função espera receber
- **Argumento** → valor real que você passa para a função
- **Retorno** → o valor que a função devolve com `return`

## 📌 3.1. Funções simples (sem parâmetros e sem retorno)

A forma mais básica: uma função que **só executa um bloco de código** quando chamada, sem receber nem retornar nada.

In [ ]:
# ============================================================
# EXEMPLO 1 - Função sem parâmetros e sem retorno
# ============================================================

# Definimos a função com 'def' seguido do nome e parênteses
# O código da função fica INDENTADO (com espaços no início)
def saudacao():
    print("=" * 30)
    print("  Bem-vindo(a) ao sistema!")
    print("=" * 30)

# Chamamos a função usando o nome seguido de parênteses
saudacao()   # Executa o bloco uma vez
saudacao()   # E outra vez...
saudacao()   # Sem precisar reescrever o código!

# Imagine se não tivéssemos função: teríamos que repetir
# os 3 prints toda vez que quiséssemos mostrar a saudação

  Bem-vindo(a) ao sistema!
  Bem-vindo(a) ao sistema!
  Bem-vindo(a) ao sistema!


In [ ]:
# ============================================================
# EXEMPLO 2 - Função que organiza o programa em blocos
# ============================================================

# Funções são ótimas para ORGANIZAR código
# Cada função tem UMA responsabilidade bem clara

def mostrar_cabecalho():
    print("╔══════════════════════════════╗")
    print("║     LOJA DO ZÉ - 2025        ║")
    print("╚══════════════════════════════╝")

def mostrar_menu():
    print("\nO que deseja fazer?")
    print("  1 - Ver produtos")
    print("  2 - Fazer pedido")
    print("  3 - Sair")

def mostrar_rodape():
    print("\n-- Obrigado pela visita! --")

# O programa principal fica MUITO mais limpo
mostrar_cabecalho()
mostrar_menu()
mostrar_rodape()

╔══════════════════════════════╗
║     LOJA DO ZÉ - 2025        ║
╚══════════════════════════════╝

O que deseja fazer?
  1 - Ver produtos
  2 - Fazer pedido
  3 - Sair

-- Obrigado pela visita! --


In [ ]:
# ============================================================
# EXEMPLO 3 - Função que faz cálculos internos
# ============================================================

# A função pode ter lógica INTERNA complexa
# Para quem chama, basta saber o NOME e o QUE ela faz

def tabuada_do_7():
    print("📊 Tabuada do 7:")
    # for dentro da função funciona normalmente
    for i in range(1, 11):
        resultado = 7 * i
        print(f"7 x {i:2d} = {resultado}")

# Só chamando uma vez já dispara a tabuada inteira
tabuada_do_7()

📊 Tabuada do 7:
7 x  1 = 7
7 x  2 = 14
7 x  3 = 21
7 x  4 = 28
7 x  5 = 35
7 x  6 = 42
7 x  7 = 49
7 x  8 = 56
7 x  9 = 63
7 x 10 = 70


In [ ]:
# ============================================================
# EXEMPLO 4 - Múltiplas funções trabalhando juntas
# ============================================================

# Uma função pode CHAMAR outra função!
# Isso permite construir programas mais complexos

def linha_separadora():
    print("-" * 40)

def cabecalho_recibo():
    linha_separadora()     # Chama a outra função
    print("        RECIBO DE COMPRA")
    linha_separadora()

def rodape_recibo():
    linha_separadora()
    print("   Volte sempre! :)")
    linha_separadora()

def gerar_recibo_exemplo():
    cabecalho_recibo()
    print("1x Camiseta ........ R$ 49,90")
    print("1x Calça ........... R$ 89,90")
    print("Total: ............. R$ 139,80")
    rodape_recibo()

# Basta chamar a função principal!
gerar_recibo_exemplo()

----------------------------------------
        RECIBO DE COMPRA
----------------------------------------
1x Camiseta ........ R$ 49,90
1x Calça ........... R$ 89,90
Total: ............. R$ 139,80
----------------------------------------
   Volte sempre! :)
----------------------------------------


### ✏️ Exercício 3.1 — Funções simples

Crie três funções:

1. `mostrar_linha()` — imprime uma linha com 40 sinais de igual (`=`).
2. `mostrar_titulo()` — imprime uma linha, o texto `"MEU PROGRAMA EM PYTHON"` centralizado, e outra linha.
3. `mostrar_frase_motivacional()` — imprime **3 frases** motivacionais em linhas separadas (invente as frases).

Depois, **chame as três funções** na ordem para formar um "banner" completo.

In [ ]:
# 👉 Escreva seu código aqui:

# 1. Função mostrar_linha()
def mostrar_linha():
    print("=" * 40)

# 2. Função mostrar_titulo()
def mostrar_titulo():
    mostrar_linha()
    print("  MEU PROGRAMA EM PYTHON".center(40))
    mostrar_linha()

# 3. Função mostrar_frase_motivacional()
def mostrar_frase_motivacional():
    print("\n=== Acredite em si mesmo e você será imparável. ===")
    print("\n=== O segredo para sair na frente é começar. ===")
    print("\n=== A persistência é o caminho do êxito. ===")

# Chame as funções em ordem
def gerar_banner():
    mostrar_titulo()
    mostrar_frase_motivacional()

# Chamando a função principal
gerar_banner()


          MEU PROGRAMA EM PYTHON        

=== Acredite em si mesmo e você será imparável. ===

=== O segredo para sair na frente é começar. ===

=== A persistência é o caminho do êxito. ===


## 📌 3.2. Funções com parâmetros

**Parâmetros** permitem que a função receba **dados de entrada** para trabalhar. Isso deixa a função MUITO mais útil e flexível!

```python
def cumprimentar(nome):          # ← 'nome' é o PARÂMETRO
    print(f"Olá, {nome}!")

cumprimentar("Ana")              # ← "Ana" é o ARGUMENTO
cumprimentar("Bruno")
```

In [ ]:
# ============================================================
# EXEMPLO 1 - Função com UM parâmetro
# ============================================================

# A função recebe um valor pelos parênteses e usa internamente
def saudacao_personalizada(nome):
    print(f"Olá, {nome}! Seja bem-vindo(a)!")

# Chamando a função com valores DIFERENTES
saudacao_personalizada("Ana")
saudacao_personalizada("Bruno")
saudacao_personalizada("Carlos")

# Podemos até passar o valor vindo de uma variável
usuario = "Diana"
saudacao_personalizada(usuario)

Olá, Ana! Seja bem-vindo(a)!
Olá, Bruno! Seja bem-vindo(a)!
Olá, Carlos! Seja bem-vindo(a)!
Olá, Diana! Seja bem-vindo(a)!


In [ ]:
# ============================================================
# EXEMPLO 2 - Função com MÚLTIPLOS parâmetros
# ============================================================

# Uma função pode receber vários parâmetros, separados por vírgula
def apresentar_pessoa(nome, idade, cidade):
    print(f"Nome: {nome}")
    print(f"Idade: {idade} anos")
    print(f"Cidade: {cidade}")
    print("-" * 30)

# Passamos os argumentos na MESMA ORDEM dos parâmetros
apresentar_pessoa("Ana", 25, "Recife")
apresentar_pessoa("Bruno", 30, "São Paulo")

# Também podemos passar por NOME (não importa a ordem)
# Isso se chama 'argumentos nomeados' ou 'keyword arguments'
apresentar_pessoa(cidade="Curitiba", nome="Carlos", idade=28)

Nome: Ana
Idade: 25 anos
Cidade: Recife
------------------------------
Nome: Bruno
Idade: 30 anos
Cidade: São Paulo
------------------------------
Nome: Carlos
Idade: 28 anos
Cidade: Curitiba
------------------------------


In [ ]:
# ============================================================
# EXEMPLO 3 - Parâmetros com VALOR PADRÃO
# ============================================================

# Podemos definir um VALOR PADRÃO para o parâmetro
# Se a pessoa não passar esse argumento, ele assume o padrão

def cumprimento(nome, horario="dia"):
    # Se 'horario' não for passado, usa "dia" como padrão
    print(f"Bom {horario}, {nome}!")

# Forma 1: passando os dois argumentos
cumprimento("Ana", "manhã")   # Bom manhã, Ana!
cumprimento("Bruno", "tarde") # Bom tarde, Bruno!
cumprimento("Carla", "noite") # Bom noite, Carla!

# Forma 2: omitindo o segundo (usa o padrão)
cumprimento("Diana")          # Bom dia, Diana! (padrão)

print("\n---\n")

# Exemplo prático: função de desconto
def calcular_preco(preco_original, desconto=0):
    # Se o desconto não for passado, assume 0 (sem desconto)
    preco_final = preco_original - (preco_original * desconto / 100)
    print(f"Preço: R$ {preco_final:.2f} (desconto de {desconto}%)")

calcular_preco(100)           # Sem desconto
calcular_preco(100, 10)       # 10% de desconto
calcular_preco(100, 50)       # 50% de desconto

Bom manhã, Ana!
Bom tarde, Bruno!
Bom noite, Carla!
Bom dia, Diana!

---

Preço: R$ 100.00 (desconto de 0%)
Preço: R$ 90.00 (desconto de 10%)
Preço: R$ 50.00 (desconto de 50%)


In [ ]:
# ============================================================
# EXEMPLO 4 - Função recebendo uma lista como parâmetro
# ============================================================

# Parâmetros podem ser de QUALQUER tipo: número, texto, lista, dicionário...

def mostrar_lista_compras(itens):
    # itens é uma lista, tratamos normalmente
    print("🛒 Lista de compras:")
    for i, item in enumerate(itens, start=1):
        print(f"  {i}. {item}")
    print(f"Total de itens: {len(itens)}")

# Passando diferentes listas
compras_semana = ["pão", "leite", "ovos", "café", "açúcar"]
mostrar_lista_compras(compras_semana)

print()

compras_festa = ["refrigerante", "salgadinho", "bolo"]
mostrar_lista_compras(compras_festa)

# Também podemos passar a lista direto, sem variável
mostrar_lista_compras(["frutas", "verduras"])

🛒 Lista de compras:
  1. pão
  2. leite
  3. ovos
  4. café
  5. açúcar
Total de itens: 5

🛒 Lista de compras:
  1. refrigerante
  2. salgadinho
  3. bolo
Total de itens: 3
🛒 Lista de compras:
  1. frutas
  2. verduras
Total de itens: 2


### ✏️ Exercício 3.2 — Funções com parâmetros

Crie uma função `cadastrar_aluno(nome, idade, curso="Python")` que:

1. Receba os parâmetros `nome`, `idade` e `curso` (com valor padrão `"Python"`).
2. Imprima os dados formatados assim:

```
=== Cadastro ===
Nome: Ana
Idade: 22 anos
Curso: Python
```

Depois chame a função **três vezes**:
1. Com nome e idade apenas (usando o curso padrão).
2. Com nome, idade e curso `"JavaScript"`.
3. Usando argumentos nomeados em ordem diferente.

In [ ]:
# 👉 Escreva seu código aqui:

# Definir a função
def cadastrar_aluno(nome, idade, curso="Python"):
    print("=== Cadastro ===")
    print(f"Nome: {nome}")
    print(f"Idade: {idade} anos")
    print(f"Curso: {curso}")

# Chamada 1: só nome e idade
print("\n---\n")
cadastrar_aluno("Ana", 22)

# Chamada 2: com curso JavaScript
print("\n---\n")
cadastrar_aluno("Bruno", 25, "JavaScript")

# Chamada 3: argumentos nomeados em ordem diferente
print("\n---\n")
cadastrar_aluno(idade=28, nome="Carlos", curso="Java")



---

=== Cadastro ===
Nome: Ana
Idade: 22 anos
Curso: Python

---

=== Cadastro ===
Nome: Bruno
Idade: 25 anos
Curso: JavaScript

---

=== Cadastro ===
Nome: Carlos
Idade: 28 anos
Curso: Java


## 📌 3.3. Funções com retorno (return)

A palavra-chave `return` faz a função **devolver** um valor para quem a chamou. Esse valor pode ser **guardado em uma variável** ou **usado diretamente**.

**Diferença importante:**
- `print()` → **mostra** o valor na tela (visível para o usuário)
- `return` → **devolve** o valor para ser usado no programa (visível para o código)

```python
def somar(a, b):
    return a + b     # devolve o resultado

resultado = somar(3, 5)   # guarda o 8 em 'resultado'
print(resultado)          # aí sim mostra na tela
```

In [ ]:
# ============================================================
# EXEMPLO 1 - Função com return simples
# ============================================================

# Função que CALCULA e DEVOLVE o resultado
def dobro(numero):
    resultado = numero * 2
    return resultado   # devolve o valor para quem chamou
    # IMPORTANTE: tudo depois do return NÃO é executado!

# O que a função retorna pode ser guardado numa variável
x = dobro(5)
print(x)   # 10

# Ou usado diretamente em uma expressão
print(dobro(10))           # 20
print(dobro(3) + dobro(4)) # 6 + 8 = 14

# Até passar o retorno para outra função!
total = dobro(dobro(5))   # dobro(5)=10, dobro(10)=20
print(total)   # 20

10
20
14
20


In [ ]:
# ============================================================
# EXEMPLO 2 - Diferença entre print e return
# ============================================================

# Função com PRINT (mostra na tela mas não devolve nada usável)
def somar_com_print(a, b):
    print(a + b)   # Só mostra, não devolve

# Função com RETURN (devolve para ser usado)
def somar_com_return(a, b):
    return a + b   # Devolve o valor

# Vamos ver a diferença na prática
print("Testando com print:")
resultado1 = somar_com_print(3, 5)   # Imprime 8
print(f"Variável capturou: {resultado1}")   # None! (não retornou nada)

print("\nTestando com return:")
resultado2 = somar_com_return(3, 5)  # Nada aparece...
print(f"Variável capturou: {resultado2}")   # 8 (valor útil!)

# Com return, podemos FAZER COISAS com o resultado
dobro_soma = somar_com_return(3, 5) * 2
print(f"Dobro da soma: {dobro_soma}")   # 16

Testando com print:
8
Variável capturou: None

Testando com return:
Variável capturou: 8
Dobro da soma: 16


In [ ]:
# ============================================================
# EXEMPLO 3 - Função que calcula a média de uma lista
# ============================================================

# Função útil e reutilizável: calcula média de qualquer lista de números
def calcular_media(numeros):
    if len(numeros) == 0:   # Proteção: lista vazia daria erro de divisão por zero
        return 0
    total = sum(numeros)
    media = total / len(numeros)
    return media   # devolve a média calculada

# Usando a função com diferentes listas
notas_ana = [8.5, 9.0, 7.5, 8.0]
notas_bruno = [6.0, 7.0, 5.5, 6.5]

media_ana = calcular_media(notas_ana)
media_bruno = calcular_media(notas_bruno)

print(f"Média da Ana: {media_ana:.2f}")
print(f"Média do Bruno: {media_bruno:.2f}")

# O poder das funções: podemos COMPARAR os retornos facilmente!
if media_ana > media_bruno:
    print("🏆 Ana tirou a maior média!")
else:
    print("🏆 Bruno tirou a maior média!")

Média da Ana: 8.25
Média do Bruno: 6.25
🏆 Ana tirou a maior média!


In [ ]:
# ============================================================
# EXEMPLO 4 - Função retornando MÚLTIPLOS valores
# ============================================================

# Em Python, uma função pode retornar VÁRIOS valores de uma vez!
# Separa com vírgula no return e recebe com vírgula na chamada

def estatisticas(numeros):
    # Calcula várias estatísticas e devolve todas
    maior = max(numeros)
    menor = min(numeros)
    soma = sum(numeros)
    media = soma / len(numeros)

    # Retornando 4 valores de uma vez
    return maior, menor, soma, media

# Recebendo cada valor em uma variável (na mesma ordem!)
notas = [7.5, 4.0, 9.2, 6.8, 10.0, 3.5, 8.0]
mx, mn, s, m = estatisticas(notas)

print(f"📊 Estatísticas das notas:")
print(f"  Maior nota:  {mx}")
print(f"  Menor nota:  {mn}")
print(f"  Soma total:  {s:.2f}")
print(f"  Média:       {m:.2f}")

# Também podemos usar return com if para criar FLUXO
def verificar_idade(idade):
    if idade < 18:
        return "Menor de idade"
    elif idade < 60:
        return "Adulto"
    else:
        return "Idoso"

print(f"\n{verificar_idade(15)}")   # Menor de idade
print(verificar_idade(30))           # Adulto
print(verificar_idade(70))           # Idoso

📊 Estatísticas das notas:
  Maior nota:  10.0
  Menor nota:  3.5
  Soma total:  49.00
  Média:       7.00

Menor de idade
Adulto
Idoso


### ✏️ Exercício 3.3 — Funções com retorno

Crie as seguintes funções, **todas com `return`**:

1. `calcular_imc(peso, altura)` → retorna o IMC (fórmula: `peso / (altura ** 2)`).
2. `classificar_imc(imc)` → recebe um valor de IMC e retorna a classificação:
   - Abaixo de 18.5: `"Abaixo do peso"`
   - Entre 18.5 e 24.9: `"Peso normal"`
   - Entre 25.0 e 29.9: `"Sobrepeso"`
   - 30.0 ou mais: `"Obesidade"`
3. No programa principal:
   - Peça peso e altura ao usuário com `input()` (converta para `float`).
   - Chame `calcular_imc()` para obter o IMC.
   - Chame `classificar_imc()` passando o resultado anterior.
   - Mostre o IMC (com 2 casas decimais) e a classificação.

In [ ]:
# 👉 Escreva seu código aqui:

# 1. Função calcular_imc
def calcular_imc(peso, altura):
    imc = peso / (altura ** 2)
    return round(imc, 2)

# 2. Função classificar_imc
def classificar_imc(imc):
    if imc < 18.5:
        return "Abaixo do peso"
    elif imc < 25:
        return "Peso normal"
    elif imc < 30:
        return "Sobrepeso"
    else:
        return "Obesidade"

# 3. Programa principal
input_peso = input("Digite seu peso (em kg): ")
input_altura = input("Digite sua altura (em metros): ")

peso = float(input_peso)
altura = float(input_altura.replace(',', '.')) # Corrigido: substitui vírgula por ponto

imc = calcular_imc(peso, altura)
classificacao = classificar_imc(imc) # Corrigido: atribui o resultado da função

print(f"\nSeu IMC é: {imc:.2f}")
print(f"Classificação: {classificacao}")

Digite seu peso (em kg): 100
Digite sua altura (em metros): 1,73

Seu IMC é: 33.41
Classificação: Obesidade


## 📌 3.4. Escopo de variáveis (local vs global)

Toda variável criada **dentro de uma função** existe **apenas ali dentro** — é chamada de **variável local**.

Variáveis criadas **fora** de qualquer função são **globais** — podem ser lidas em qualquer lugar (mas alterar elas dentro de função exige a palavra `global`).

Esse conceito é **FUNDAMENTAL** para evitar bugs!

In [ ]:
# ============================================================
# EXEMPLO 1 - Variável LOCAL (dentro da função)
# ============================================================

def minha_funcao():
    # 'x' é uma variável LOCAL — só existe dentro desta função
    x = 10
    print(f"Dentro da função, x = {x}")

# Chamando a função
minha_funcao()

# Tentar acessar 'x' aqui fora dá ERRO!
# O Python diz: NameError: name 'x' is not defined
try:
    print(x)   # Vai dar erro
except NameError:
    print("❌ Erro! 'x' não existe fora da função.")

Dentro da função, x = 10
10


In [ ]:
# ============================================================
# EXEMPLO 2 - Variável GLOBAL (fora da função)
# ============================================================

# 'nome_site' foi criada FORA de qualquer função → é GLOBAL
nome_site = "Minha Loja"

def mostrar_cabecalho():
    # Podemos LER a variável global normalmente
    print(f"=== {nome_site} ===")

def mostrar_rodape():
    # Também conseguimos ler em outra função
    print(f"© {nome_site} - Todos os direitos reservados")

mostrar_cabecalho()
print("Conteúdo do site...")
mostrar_rodape()

# IMPORTANTE: ler está OK. Modificar é diferente! (próximo exemplo)

=== Minha Loja ===
Conteúdo do site...
© Minha Loja - Todos os direitos reservados


In [ ]:
# ============================================================
# EXEMPLO 3 - A 'armadilha' ao tentar modificar global
# ============================================================

contador = 0   # Variável global

def incrementar_errado():
    # Isso cria uma nova variável LOCAL 'contador' (não mexe na global!)
    contador = contador + 1   # ❌ ERRO! não consegue ler e escrever ao mesmo tempo

# O correto: usar a palavra 'global' para dizer 'quero mexer na variável global'
def incrementar_certo():
    global contador   # 'estou falando da variável global contador'
    contador = contador + 1

print(f"Antes: {contador}")
incrementar_certo()
incrementar_certo()
incrementar_certo()
print(f"Depois de 3 chamadas: {contador}")   # 3

# ATENÇÃO: usar 'global' é RARO na prática.
# A forma mais limpa e recomendada é usar PARÂMETROS e RETURN.
# Veja o próximo exemplo para entender.

Antes: 0
Depois de 3 chamadas: 3


In [ ]:
# ============================================================
# EXEMPLO 4 - Forma correta: parâmetros e retornos
# ============================================================

# Em vez de mexer em variáveis globais, PASSE valores como parâmetro
# e DEVOLVA o resultado com return. Isso deixa o código muito mais claro.

# ❌ Forma problemática (usa global)
saldo = 1000
def sacar_ruim(valor):
    global saldo
    saldo -= valor

# ✅ Forma limpa (usa parâmetros e return)
def sacar_bom(saldo_atual, valor):
    # Não mexe em variável global, só usa o que recebe e devolve o novo valor
    novo_saldo = saldo_atual - valor
    return novo_saldo

# Uso da função correta
meu_saldo = 1000
print(f"Saldo inicial: R$ {meu_saldo}")

meu_saldo = sacar_bom(meu_saldo, 250)
print(f"Após sacar R$ 250: R$ {meu_saldo}")

meu_saldo = sacar_bom(meu_saldo, 100)
print(f"Após sacar R$ 100: R$ {meu_saldo}")

# Vantagens desta forma:
# - A função é 'pura': dado os mesmos parâmetros, sempre retorna o mesmo resultado
# - Fica mais fácil testar e encontrar bugs
# - Dá para reaproveitar com qualquer variável

Saldo inicial: R$ 1000
Após sacar R$ 250: R$ 750
Após sacar R$ 100: R$ 650


### ✏️ Exercício 3.4 — Escopo de variáveis

Crie uma função `aplicar_desconto(preco, percentual)` que:

1. Recebe um preço e um percentual de desconto.
2. Cria uma variável LOCAL `valor_desconto = preco * percentual / 100`.
3. Cria uma variável LOCAL `preco_final = preco - valor_desconto`.
4. Retorna `preco_final`.

Depois, no programa principal:
1. Crie uma variável (fora da função) com um preço, por exemplo `preco_produto = 200`.
2. Chame a função passando 10% de desconto e guarde o retorno em `novo_preco`.
3. Tente imprimir `valor_desconto` fora da função e explique com um comentário por que dá erro.
4. Mostre o preço original e o preço com desconto.

In [ ]:
# 👉 Escreva seu código aqui:

# 1. Função aplicar_desconto
def aplicar_desconto(preco, percentual):
    valor_desconto = preco * percentual / 100
    preco_final = preco - valor_desconto
    return preco_final

# 2. Variável global com o preço
local_preco_produto = 200
novo_preco = aplicar_desconto(local_preco_produto, 10)

# 3. Chamar função e guardar retorno
print(f"Preço original: R$ {local_preco_produto}")
print(f"Preço com desconto: R$ {novo_preco}")


# 4. Tentar imprimir valor_desconto (vai dar erro — comente o motivo)
# try:
#     print(valor_desconto)
# except NameError:
#     print("Dá erro porque...")

try:
    print(f"Valor do desconto: R$ {valor_desconto}")
except NameError:
    print("Dá erro porque...")

# 5. Mostrar preço original e com desconto
print(f"Preço original: R$ {local_preco_produto}")
print(f"Preço com desconto: R$ {novo_preco}")


Preço original: R$ 200
Preço com desconto: R$ 180.0
Dá erro porque...
Preço original: R$ 200
Preço com desconto: R$ 180.0


## 📌 3.5. Funções combinando tudo (com listas, dicionários e loops)

Agora vamos **juntar tudo**: funções que recebem listas/dicionários, percorrem com loops, aplicam condicionais e retornam resultados úteis. É assim que se escreve Python de verdade!

In [ ]:
# ============================================================
# EXEMPLO 1 - Função que filtra uma lista
# ============================================================

# Função que recebe uma lista de notas e retorna só as aprovadas
def filtrar_aprovados(notas, nota_minima=6.0):
    # Cria uma nova lista vazia que vai receber só as notas >= minima
    aprovados = []
    # Percorre todas as notas recebidas
    for nota in notas:
        if nota >= nota_minima:
            aprovados.append(nota)
    return aprovados

# Testando com diferentes turmas
turma_a = [7.5, 4.0, 9.2, 6.8, 3.5, 8.0, 5.5, 10.0]
turma_b = [5.0, 6.5, 7.0, 4.5, 8.5, 9.0]

aprovados_a = filtrar_aprovados(turma_a)
aprovados_b = filtrar_aprovados(turma_b, nota_minima=7.0)   # nota mais exigente

print(f"Turma A — aprovados (>=6.0): {aprovados_a}")
print(f"Turma B — aprovados (>=7.0): {aprovados_b}")

Turma A — aprovados (>=6.0): [7.5, 9.2, 6.8, 8.0, 10.0]
Turma B — aprovados (>=7.0): [7.0, 8.5, 9.0]


In [ ]:
# ============================================================
# EXEMPLO 2 - Função que retorna um dicionário com estatísticas
# ============================================================

# Ao invés de retornar vários valores soltos, retornamos um dicionário
# organizado — fica muito mais fácil de usar

def analisar_vendas(vendas):
    return {
        "total": sum(vendas),
        "media": sum(vendas) / len(vendas),
        "maior": max(vendas),
        "menor": min(vendas),
        "quantidade": len(vendas)
    }

# Vendas da semana
vendas_semana = [1200, 850, 1500, 2300, 980, 1100, 1800]

# Chamando a função — recebemos um dicionário completo de volta
resultado = analisar_vendas(vendas_semana)

# Acessando os dados com as chaves
print("📊 RELATÓRIO DE VENDAS DA SEMANA")
print("=" * 40)
print(f"Quantidade de vendas: {resultado['quantidade']}")
print(f"Total vendido:        R$ {resultado['total']:.2f}")
print(f"Média diária:         R$ {resultado['media']:.2f}")
print(f"Maior venda:          R$ {resultado['maior']:.2f}")
print(f"Menor venda:          R$ {resultado['menor']:.2f}")

📊 RELATÓRIO DE VENDAS DA SEMANA
Quantidade de vendas: 7
Total vendido:        R$ 9730.00
Média diária:         R$ 1390.00
Maior venda:          R$ 2300.00
Menor venda:          R$ 850.00


In [ ]:
# ============================================================
# EXEMPLO 3 - Função que trabalha com dicionário
# ============================================================

# Função que recebe um dicionário de produtos e aplica um reajuste
def reajustar_precos(produtos, percentual):
    # Criamos um NOVO dicionário com os preços atualizados
    produtos_novos = {}
    for nome, preco in produtos.items():
        novo_preco = preco + (preco * percentual / 100)
        produtos_novos[nome] = round(novo_preco, 2)   # round arredonda para 2 casas
    return produtos_novos

# Dicionário de produtos
cardapio = {
    "hamburguer": 25.00,
    "pizza": 45.00,
    "refrigerante": 8.00,
    "sobremesa": 15.00
}

# Aplicando reajuste de 10%
novo_cardapio = reajustar_precos(cardapio, 10)

# Comparando antigo vs novo
print("📋 REAJUSTE DE PREÇOS (+10%)")
print("=" * 45)
print(f"{'Produto':<15} {'Antes':>10} {'Depois':>10}")
print("-" * 45)
# Percorremos as chaves e mostramos os dois preços
for produto in cardapio:
    antigo = cardapio[produto]
    novo = novo_cardapio[produto]
    print(f"{produto:<15} R$ {antigo:>6.2f} R$ {novo:>6.2f}")

📋 REAJUSTE DE PREÇOS (+10%)
Produto              Antes     Depois
---------------------------------------------
hamburguer      R$  25.00 R$  27.50
pizza           R$  45.00 R$  49.50
refrigerante    R$   8.00 R$   8.80
sobremesa       R$  15.00 R$  16.50


In [ ]:
# ============================================================
# EXEMPLO 4 - Sistema completo com várias funções
# ============================================================

# Um 'mini-sistema' usando várias funções que se complementam
# Simulando um sistema de cadastro de alunos com cálculo de média

def criar_aluno(nome, notas):
    # Retorna um dicionário representando o aluno
    return {
        "nome": nome,
        "notas": notas,
        "media": calcular_media_aluno(notas),
        "situacao": classificar(calcular_media_aluno(notas))
    }

def calcular_media_aluno(notas):
    return sum(notas) / len(notas)

def classificar(media):
    if media >= 7.0:
        return "Aprovado"
    elif media >= 5.0:
        return "Recuperação"
    else:
        return "Reprovado"

def mostrar_aluno(aluno):
    print(f"👤 {aluno['nome']}")
    print(f"   Notas: {aluno['notas']}")
    print(f"   Média: {aluno['media']:.2f}")
    print(f"   Situação: {aluno['situacao']}")
    print()

# PROGRAMA PRINCIPAL
# Criando uma lista de alunos usando as funções
turma = [
    criar_aluno("Ana",     [8.5, 9.0, 7.5, 8.0]),
    criar_aluno("Bruno",   [6.0, 5.5, 5.0, 6.5]),
    criar_aluno("Carlos",  [3.0, 4.5, 2.0, 4.0]),
    criar_aluno("Diana",   [9.5, 10.0, 9.0, 9.2])
]

# Exibindo cada aluno
print("📚 RELATÓRIO DA TURMA")
print("=" * 40)
for aluno in turma:
    mostrar_aluno(aluno)

# Estatísticas finais
aprovados = sum(1 for a in turma if a["situacao"] == "Aprovado")
print(f"Total de aprovados: {aprovados}/{len(turma)}")

📚 RELATÓRIO DA TURMA
👤 Ana
   Notas: [8.5, 9.0, 7.5, 8.0]
   Média: 8.25
   Situação: Aprovado

👤 Bruno
   Notas: [6.0, 5.5, 5.0, 6.5]
   Média: 5.75
   Situação: Recuperação

👤 Carlos
   Notas: [3.0, 4.5, 2.0, 4.0]
   Média: 3.38
   Situação: Reprovado

👤 Diana
   Notas: [9.5, 10.0, 9.0, 9.2]
   Média: 9.43
   Situação: Aprovado

Total de aprovados: 2/4


### ✏️ Exercício 3.5 — Funções combinando tudo

Crie um mini-sistema para **registrar e analisar temperaturas** da semana:

1. Crie uma função `criar_registro(dia, temperatura)` que retorna um dicionário com as chaves `"dia"` e `"temperatura"`.

2. Crie uma função `temperatura_mais_alta(registros)` que recebe uma lista de registros e retorna o **dicionário completo** do dia mais quente.

3. Crie uma função `media_temperaturas(registros)` que retorna a média das temperaturas.

4. Crie uma função `dias_acima_da_media(registros)` que retorna uma **lista dos dias** (só o nome do dia) cujas temperaturas ficaram **acima da média**.

5. No programa principal:
   - Crie uma lista `semana` com pelo menos 5 registros usando `criar_registro()`.
   - Mostre a temperatura mais alta (dia e valor).
   - Mostre a média (com 2 casas decimais).
   - Mostre quais dias ficaram acima da média.

In [ ]:
from IPython.core.display import set_matplotlib_close
# 👉 Escreva seu código aqui:

# 1. Função criar_registro
def criar_registro(dia, temperatura):
    return {"dia": dia, "temperatura": temperatura}

# 2. Função temperatura_mais_alta
def temperatura_mais_alta(registros):
    mais_alta = registros[0]
    for registro in registros:
        if registro["temperatura"] > mais_alta["temperatura"]:
            mais_alta = registro
    return mais_alta

# 3. Função media_temperaturas
def media_temperaturas(registros):
    soma = 0
    for registro in registros:
        soma += registro["temperatura"]
    return soma / len(registros)

# 4. Função dias_acima_da_media
def dias_acima_da_media(registros):
    media = media_temperaturas(registros)
    dias = []
    for registro in registros:
        if registro["temperatura"] > media:
            dias.append(registro["dia"])
    return dias

# 5. Programa principal
semana = [
    criar_registro("Segunda", 28.5),
    criar_registro("Terça", 30.0),
    criar_registro("Quarta", 29.5),
    criar_registro("Quinta", 27.8),
    criar_registro("Sexta", 29.2)
]

mais_alta = temperatura_mais_alta(semana)
print(f"Temperatura mais alta: {mais_alta['dia']} - {mais_alta['temperatura']}°C")

media = media_temperaturas(semana)
print(f"Média das temperaturas: {media:.2f}°C")

dias_acima_media = dias_acima_da_media(semana)

print("Dias com temperatura acima da média:")
for dia in dias_acima_media:
    print(dia)


Temperatura mais alta: Terça - 30.0°C
Média das temperaturas: 29.00°C
Dias com temperatura acima da média:
Terça
Quarta
Sexta


---
# 🎉 Parabéns! Você concluiu a Parte 2!

**Nesta aula você aprendeu:**

### 📦 Listas
- Criar, acessar (índice e índice negativo), modificar
- Fatiamento (`lista[a:b]`)
- Adicionar: `.append()`, `.insert()`, `.extend()`
- Remover: `.remove()`, `.pop()`, `del`, `.clear()`
- Buscar: `in`, `.index()`, `.count()`
- Funções úteis: `len()`, `max()`, `min()`, `sum()`
- Ordenar: `.sort()`, `sorted()`, `.reverse()`

### 📚 Dicionários
- Estrutura `chave: valor`
- Acesso seguro com `.get()`
- Percorrer com `.keys()`, `.values()`, `.items()`
- Adicionar, modificar e remover pares
- Dicionários aninhados (dicionário dentro de dicionário ou dentro de lista)

### 🔧 Funções
- Criar funções com `def`
- Parâmetros e argumentos (incluindo valores padrão)
- Retorno com `return` (um ou múltiplos valores)
- Escopo local vs global
- Combinar funções com listas, dicionários e condicionais

## 🚀 Próximos passos
- Pratique refatorando programas antigos, transformando pedaços repetidos em funções.
- Tente criar um pequeno sistema (cadastro, jogo, calculadora) usando tudo que aprendeu.
- Explore **tratamento de erros** com `try/except` e **leitura/escrita de arquivos** nas próximas aulas.

**Bons estudos! 🐍💻**